src/main/java/algorithms/curated170/medium/TheMaze.java

```diff
@@ -0,0 +1,66 @@
+package algorithms.curated170.medium;
+
+public class TheMaze {
+
+    public static void main(String[] args) {
+        TheMaze theMaze = new TheMaze();
+        int[][] maze = {{0, 0, 1, 0, 0},
+        {0, 0, 0, 0, 0},
+        {0, 0, 0, 1, 0},
+        {1, 1, 0, 1, 1},
+        {0, 0, 0, 0, 0}};
+
+        int[] start = {0, 4}, destination = {4, 4};
+        boolean result = theMaze.hasPath(maze, start, destination);
+        System.out.println(result);
+    }
+    boolean visited[][];
+    int[][] directions = {{0, -1}, {1, 0}, {-1, 0}, {0, 1}};
+    int[][] maze;
+    int[] destination;
+
+    public boolean hasPath(int[][] maze, int[] start, int[] destination) {
+        visited = new boolean[maze.length][maze[0].length];
+        this.maze = maze;
+        this.destination = destination;
+
+        return dfs(start[0], start[1]);
+    }
+
+    public boolean dfs(int x, int y) {
+
+        if (x == destination[0] && y == destination[1]) {
+            return true;
+        }
+
+        visited[x][y] = true;
+        boolean result = false;
+
+        for (int i = 0; i < directions.length; i++) {
+            int dirx = directions[i][0];
+            int diry = directions[i][1];
+            int nextX = x + dirx;
+            int nextY = y + diry;
+
+            while (canPass(nextX, nextY)
+                    && !visited[nextX][nextY]) {
+                int prx = nextX;
+                int pry = nextY;
+                nextX += dirx;
+                nextY += diry;
+                result = result || (!canPass(nextX, nextY) && dfs(prx, pry));
```

Comment:
Let's say that we have went left from some tile. This loop implies that we will also traverse the same from left to right until we see that the right tile is already visited. Isn't this to some degree duplicate unneeded calculation or am I seeing it wrong?

Reply:
It is duplicated but I am not sure if it is unnecessary.

Somehow we have to distinguish the above-mentioned case and the case in Example-2.

It is one of the alternative ways.

We can change it when we find a more efficient way.

Reply:
We can perhaps put a simple statement like "Don't go right" to the method call.  We can wrap this up in an integer or a boolean value.